File: leetcode/283-move-zeroes/solution.py
from typing import List


# LeetCode 283: Move Zeroes
class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        insert_pos = 0
        for num in nums:
            if num != 0:
                nums[insert_pos] = num
                insert_pos += 1
        for i in range(insert_pos, len(nums)):
            nums[i] = 0


if __name__ == "__main__":
    nums = [0, 1, 0, 3, 12]
    Solution().moveZeroes(nums)
    print(nums)  # [1, 3, 12, 0, 0]

File: leetcode/283-move-zeroes/solution.cpp
#include <bits/stdc++.h>
using namespace std;

// LeetCode 283: Move Zeroes
class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        int insertPos = 0;
        for (int num : nums) {
            if (num != 0) {
                nums[insertPos++] = num;
            }
        }
        while (insertPos < (int)nums.size()) {
            nums[insertPos++] = 0;
        }
    }
};

int main() {
    Solution sol;
    vector<int> nums = {0, 1, 0, 3, 12};
    sol.moveZeroes(nums);
    for (int num : nums) cout << num << " "; // 1 3 12 0 0
    cout << endl;
    return 0;
}
